Source/core/FileSystem.cpp

```diff
@@ -56,6 +56,90 @@ namespace Core {
     {
         Close();
     }
+
+    /* static */ string File::Normalize(const string& location, bool& valid)
+    {
+        string result(location);
+
+        valid = true;
+
+        // First see if we are not empy.
+        if (result.empty() == false) {
+            uint32_t index = 0;
+
+           while (index < result.length()) {
+
+#ifdef __WINDOWS__
+                if (result[index] == '\\') {
+                    result[index] = '/';
+                }
+#endif
+
+                if ((result[index] == '/') && (index >= 1) ) {
+                    if (result[index - 1] == '/') {
+                        if (index >= 2) {
+                            // We have a double slash, clear all till the beginning
+                           result = result.substr(index - 1);
+                           index = 1;
+                        }
+                    }
+                    else if (result[index - 1] == '.') {
+                        if ((index == 1) || (result[index - 2] == '/')) {
+                            // It is a dot, remove it...
+                            uint32_t offset = (index == 1 ? 0 : index - 2);
+                            result.erase(offset, 2);
+                            index = offset;
+                        }
+                        else if ((result[index - 2] == '.') && ((index == 2) || (result[index - 3] == '/'))) {
+                            if (index <= 3) {
+                                valid = false;
+                                result.clear();
+                            }
+                            else {
+                                // Seems like we are moving up a directory... execute that on the result... if we can...
+                                // there is data we can drop, drop it, drop it till the '/' is found
+                                uint32_t offset = (index <= 3 ? 0 : index - 4);
+                                while ((offset > 0) && (result[offset] != '/')) {
+                                    offset--;
+                                }
+                                result.erase(offset, index - offset);
```

Comment:
Think something might be wrong here in case you have something like "aaa/../bbbb" in that case if I'm right (doing this from my head I'm not a compiler/debugger :) ) the result would be "/bbbb". I understand that on any other then the first folder you want to keep the/ but when it is not present in the original I guess you do not want to introduce it (or perhaps it is required that is always should start with a / but the better check already at the start of the method)